File: benchmarks/pandas/bench_holiday_observances.py
"""
Benchmark: Holiday observance functions from pandas.tseries.holiday.

Compares against tsb (TypeScript) bench_holiday_observances.ts.
"""
import datetime
import json
import time

from pandas.tseries.holiday import (
    nearest_workday,
    next_monday,
    next_monday_or_tuesday,
    previous_friday,
    previous_workday,
    sunday_to_monday,
)

N = 1_000
WARMUP = 5
ITERS = 50

base = datetime.date(2000, 1, 1)
dates = [base + datetime.timedelta(days=i) for i in range(N)]

# warm-up
for _ in range(WARMUP):
    for d in dates:
        nearest_workday(d)
        next_monday(d)
        next_monday_or_tuesday(d)
        previous_friday(d)
        previous_workday(d)
        sunday_to_monday(d)

t0 = time.perf_counter()
for _ in range(ITERS):
    for d in dates:
        nearest_workday(d)
        next_monday(d)
        next_monday_or_tuesday(d)
        previous_friday(d)
        previous_workday(d)
        sunday_to_monday(d)
total_ms = (time.perf_counter() - t0) * 1000

print(
    json.dumps(
        {
            "function": "holiday_observances",
            "mean_ms": total_ms / ITERS,
            "iterations": ITERS,
            "total_ms": total_ms,
        }
    )
)

File: benchmarks/tsb/bench_holiday_observances.ts
/**
 * Benchmark: Holiday observance functions (nearestWorkday, nextMonday, previousFriday, etc.)
 *
 * Mirrors pandas.tseries.holiday observance helpers:
 * nearest_workday, next_monday, previous_friday, sunday_to_monday, etc.
 */
import {
  nearestWorkday,
  nextMonday,
  nextMondayOrTuesday,
  previousFriday,
  previousWorkday,
  sundayToMonday,
} from "../../src/index.js";

const N = 1_000;
const WARMUP = 5;
const ITERS = 50;

// Create N dates spread across multiple years
const BASE = new Date("2000-01-01").getTime();
const MS_PER_DAY = 86_400_000;
const dates: Date[] = Array.from({ length: N }, (_, i) => new Date(BASE + i * MS_PER_DAY));

let t0 = performance.now();
for (let i = 0; i < WARMUP; i++) {
  for (const d of dates) {
    nearestWorkday(d);
    nextMonday(d);
    nextMondayOrTuesday(d);
    previousFriday(d);
    previousWorkday(d);
    sundayToMonday(d);
  }
}
t0 = performance.now();

for (let i = 0; i < ITERS; i++) {
  for (const d of dates) {
    nearestWorkday(d);
    nextMonday(d);
    nextMondayOrTuesday(d);
    previousFriday(d);
    previousWorkday(d);
    sundayToMonday(d);
  }
}
const total_ms = performance.now() - t0;

console.log(
  JSON.stringify({
    function: "holiday_observances",
    mean_ms: total_ms / ITERS,
    iterations: ITERS,
    total_ms,
  }),
);
